### PR TITLE
Add configuration file support for qmanager

### PR DESCRIPTION
**Pull request by dongahn**
This PR is work in progress as I need to populate test cases.
It incorporate configuration file support for the qmanager module.
With this support, qmanager enforces the queuing policy
and parameters with the following precedence. (A lower-numbered
config source takes precedence over a higher numbered source.)

1. module load-time parameters (which can be specified either
   manually for reload or through RC environment variables
   for the initial load);
2. configuration file load-time parameters;
3. compilation-time parameters (defined in
    qmanager/config/queue_system_defaults.hpp
     and qmanager/module/qmanager.cpp).

Propagate the queuing parameters and queuing policy
parameters to the policy subsystem by augmenting
the "apply_params" interface.

### CODE REVIEW

**Comment by dongahn** [comment]
Just added some test cases and made minor changes to solve a few issues caught during running these tests. Took off WIP as this should be reviewed now and merged after that.

File: qmanager/modules/qmanager.cpp [hunk 1]
```diff
@@ -326,6 +387,122 @@ static void qmanager_destroy (std::shared_ptr<qmanager_ctx_t> &ctx)
...
+        ctx->conf.policy_params += "reservation-depth="
+                                       + std::to_string (reservation_depth);
+    }
+    return 0;
+}
+
+static int process_config_file (std::shared_ptr<qmanager_ctx_t> &ctx)
+{
+    int rc = 0;
+    if ((rc = process_config_queue_policy (ctx)) < 0)
```
Comment (garlick) [inline qmanager/modules/qmanager.cpp]: Suggestion: add a check for TOML parse errors here and make that fatal, something like:
```c
flux_conf_error_t error;
if (!flux_get_conf (h, &error)) {
    flux_log_msg (h, "%s: %d: %s", error.filename, error.lineno, error.errbuf);
    return -1;
}
```
After that, the config object will be cached in the handle and it's safe to call `flux_get_conf (h, NULL)` without error checking to access it.

Comment (dongahn) [reply qmanager/modules/qmanager.cpp]: Good point. I will add this check.

**Comment by dongahn** [comment]
Ok.  I addressed @garlick's review above. And I also added a bunch of changes to solve the Travis failures. I forced a push. @SteVwonder: once you can take a look at this, this can go in from my perspective.

File: t/t1005-qmanager-conf.t [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    export FLUX_CONF_DIR=${conf_base}/04-conservative &&
+    flux broker flux module list | grep qmanager &&
+    unset FLUX_CONF_DIR
+'
+
+test_expect_success 'qmanager: load works with small queue-depth' '
+    export FLUX_CONF_DIR=${conf_base}/05-small-queue-depth &&
+    flux broker flux module list | grep qmanager &&
+    unset FLUX_CONF_DIR
+'
```
[the remark this answers is not in the document]
Comment (SteVwonder) [reply t/t1005-qmanager-conf.t]: Related: is there any way to determine what configuration is being used when running interactively?  I cloned this PR, compiled and ran it, and then realized I don't know of a way to easy check the configuration settings.  I don't see any messages logged to `flux dmesg`.

**Comment by dongahn** [comment]
> Thanks @dongahn for putting this PR together! Some comments below. The big one being, how do you recommend going about getting the currently loaded settings for qmanager to check to see if config file changes took effect?

Yes these are needed test! But I decided to defer this until we add `flux ion` front end tool. I didn't want to add an ad-hoc front end tool and RPC for qmanager just to add these tests. Can we defer adding this test later with an issue ticket open?

File: t/conf.d/07-hybrid2/qmanager.toml [hunk 3]
```diff
@@ -0,0 +1,16 @@
...
+
+policy = "hybrid"              # queueing policy type
+
+    [qmanager.queue-params]        # general queue parameters
+    max-queue-depth = 1000000      # max queue depth (applied to all policies)
+    queue-depth = 8192             # queue-depth (applied to all policies)
+
+    [qmanager.policy-params]       # queue policy parameters
+    max-reservation-depth = 64     # max depth for "conservative" and "hybrid"
+    reservation-depth = 128        # reservation depth for HYBRID
```
[the remark this answers is not in the document]
Comment (dongahn) [reply t/conf.d/07-hybrid2/qmanager.toml]: Yes. In this case, `reservation-depth` will be set to `max-reservation-depth`. The semantics should be similar for other parameters like `max-queue-depth` vs. `queue-depth`.

File: t/t0003-basic-install.t [hunk 4]
```diff
@@ -1,37 +0,0 @@
-#!/bin/sh
```
[the remark this answers is not in the document]
Comment (dongahn) [reply t/t0003-basic-install.t]: Yes, this test isn't needed any longer. Intended change.

File: t/t1005-qmanager-conf.t (hunk 2 above)
[the remark this answers is not in the document]
Comment (dongahn) [reply t/t1005-qmanager-conf.t]: Please see my early posting. This is definitely a deficiency. We need to circle back to this when we add `flux ion` front end.

File: t/conf.d/07-hybrid2/qmanager.toml (hunk 3 above)
[the remark this answers is not in the document]
Comment (SteVwonder) [reply t/conf.d/07-hybrid2/qmanager.toml]: Ok.  That seems like a reasonable thing to do, but there should probably be a warning logged somewhere since we are overriding what the user explicitly asked for.  I imagine this popping up when a user tries to set a `reservation-depth` at load time that is larger than the config file's `max-reservation-depth` and it appear as though the command line argument is not being respected.  When in reality, it is being parsed and processed, but just overridden by the max depth.

If you agree about the warning, we can open a separate issue on that since it isn't a behaviour tied directly to this PR.

**Comment by SteVwonder** [comment]
> Can we defer adding this test later with an issue ticket open?

I'm OK with deferring the front-end tool and better CI tests until later, but I'm hesitant to merge this PR as-is because I don't know how to spot-check that it works as intended.

What are your thoughts on having `qmanager` log to flux's dmesg the policy and parameters that it is ultimately enforcing?   Then I can interactively run with different config files and load line options and see the effects in the logs.   I don't think it would be wasted effort as that is probably useful information to log anyways.

If you have a different suggestion for how to spot-check the changes, I'm open to that as well.

**Comment by dongahn** [comment]
> What are your thoughts on having qmanager log to flux's dmesg the policy and parameters that it is ultimately enforcing?

Ah... Maybe this can be used for the time being.

**Comment by dongahn** [comment]
@SteVwonder: just added a few more commits to further check correctness using `dmesg`. Thank you for making this suggestion. `dmesg` turns out to be pretty handy to do this when we don't have `flux ion` front end.

**Comment by dongahn** [comment]
I took liberty to squash the last commit so that this can be merged.

**Comment by SteVwonder** [comment]
Thanks @dongahn.  Gave this a spin, and everything checked out until I tried a `make check` after having run `make install`.  `t1005-qmanager-conf.t` started failing.  Part of the problem is #604. After fixing that though, the root cause appears:  Installing flux-sched into the same prefix as flux-core causes the `qmanager-start` and `resource-start` rc1 scripts to run whenever `flux broker` is run and the setting of `FLUX_RC_EXTRA` in `t1005-qmanager-conf.t` causing them to run a second time:

```
2020-02-29T03:53:30.450659Z broker.err[0]: rc1: flux-module: cmb.insmod: qmanager module/service is in use
2020-02-29T03:53:30.452154Z broker.err[0]: Run level 1 Exited with non-zero status (rc=1) 0.7s
```

If I comment out the `flux module load qmanager` in either of the `qmanager-start` scripts (the installed one or the in-tree one), that error goes away and the same appears for `resource`.  After repeating the commenting out for `resource`, `make check` passes.

Unfortunately, I don't have any good suggestions on how to solve this problem other than checking if the module is already loaded in the rc1 scripts to avoid `flux module load` from failing.  I am tagging @grondo (i.e., the sharness wizard) in the hopes that he has something up his sleeve already to help us here (do the RC personalities solve this issue?).

**Comment by SteVwonder** [comment]
TIL about the `FLUX_SCHED_CO_INST` variable, which is set to "co" if flux-core's prefix == the prefix that flux-sched was configured with. Maybe we can leverage that somehow?  If it is "co", then peek in the `etc/rc1.d/` directory for installed `qmanager-start` scripts.  If the flux-sched rc scripts are already installed, use those rather than the in-tree ones?

**Comment by SteVwonder** [comment]
Another idea:  we setup the rc scripts in `t/rc` to ignore/not respect `FLUX_RESOURCE_RC_NOOP`.  If we did that, we could set that variable (and its qmanager twin) to disable the installed flux-sched rc scripts and only use the in-tree testing rc scripts.

**Comment by dongahn** [comment]
> Unfortunately, I don't have any good suggestions on how to solve this problem other than checking if the module is already loaded in the rc1 scripts to avoid flux module load from failing. I am tagging @grondo (i.e., the sharness wizard) in the hopes that he has something up his sleeve already to help us here (do the RC personalities solve this issue?).

Yeah, I noticed this problem, but I left it that way for now because I didn't know of any knob I could use to solve the issue. Besides I didn't think in-tree "make check" should work correctly when you already have your modules installed to the prefix. I saw some test fails with flux-core under this condition as well. If this is a requirement and if we want to solve this problem now, then I'm all for it of course. It should be its own PR, though, I think. 

What we need would be a mechanism by which to specify "don't execute rc1 and rc3 from the installed etc direct for certain third-party scripts. Maybe we can add an environment variable `FLUX_IGNORE_INSTALLED_RC1` or similar and if it has an RC1 script with it, we don't execute those scripts in the installed directory. If ` FLUX_EXTRA_RC` is set, this RC1 script gets executed that way.

**Comment by dongahn** [comment]
> TIL about the FLUX_SCHED_CO_INST variable, which is set to "co" if flux-core's prefix == the prefix that flux-sched was configured with. Maybe we can leverage that somehow? If it is "co", then peek in the etc/rc1.d/ directory for installed qmanager-start scripts. 

> If the flux-sched rc scripts are already installed, use those rather than the in-tree ones? Seems kinda icky.

If those scripts are installed, I think we shouldn't use it for in-tree make check because it is in-tree and you may have changed your rc scripts for your PR. I think we need a way to ignore it and I don't know how to do that without introducing a knob like what I suggested.

**Comment by dongahn** [comment]
> EDIT: PS - I don't think we should remove the t0003-basic-install.t script. I think it is going to be useful for the PR I'm going to push for #604

Well... I thought `t0003-basic-install.t` too simple to be manfully useful. If this is useful, we should probably redesign one to cover all different configurations...

**Comment by dongahn** [comment]
> Another idea: we setup the rc scripts in t/rc to ignore/not respect FLUX_RESOURCE_RC_NOOP. If we did that, we could set that variable (and its qmanager twin) to disable the installed flux-sched rc scripts and only use the in-tree testing rc scripts.

This will work. But I am a bit concerned about its side effect.

`t/rc/*` are symlinks to the actual flux-sched rc scripts so it would be difficult to make this change.

We *could* make them new RC scripts instead of symlinks. But then when the actual RC scripts get modified , we have to remember these copies need to be updated as well. 

Beside, it seems like this solution is specific to flux-sched, and my opinion is that we need to solve this for all third party projects if we can. So was my suggestion above.

**Comment by dongahn** [comment]
FWIW, you will see the same issue with PR #582 as well.

**Comment by SteVwonder** [comment]
> Maybe we can add an environment variable FLUX_IGNORE_INSTALLED_RC1 or similar and if it has an RC1 script with it, we don't execute those scripts in the installed directory. If FLUX_EXTRA_RC is set, this RC1 script gets executed that way.

Yeah. I like this solution.  Seems like it handles the weird corner case we are running into while preserving the symlink to avoid the rc scripts from getting out of sync.

> If this is a requirement and if we want to solve this problem now, then I'm all for it of course.

Yeah, I think we should shoot for `make check` to still work after a `make install`.  If it doesn't, it'll be an annoyance when developing.

> It should be its own PR, though, I think.

Sure.  That's fine with me.  Let's open an issue and continue the discussion there.

> Well... I thought t0003-basic-install.t too simple to be manfully useful

I agree, it is very simple, but it is the only thing (that I'm aware of) checking the installed rc scripts (specifically that they don't fail when modules they want to remove are already removed).  It is used in the new PR that I just put up #605.  Do you want to do a quick rebase where you drop the commit that removes t0003-basic-install.t?  Alternatively, if we are OK with some noise in the history, this PR can remove it and #605 add it back in.  If the latter, I'll pop the mergify label on now.

**Comment by dongahn** [comment]
> I agree, it is very simple, but it is the only thing (that I'm aware of) checking the installed rc scripts (specifically that they don't fail when modules they want to remove are already removed). It is used in the new PR that I just put up #605. Do you want to do a quick rebase where you drop the commit that removes t0003-basic-install.t? Alternatively, if we are OK with some noise in the history, this PR can remove it and #605 add it back in. If the latter, I'll pop the mergify label on now.

Ah... Ok this makes sense then. I didn't factor into those corner cases when I removed this test. 

I'm lazy tonight... so... I like this suggestion of yours :-)

>  Alternatively, if we are OK with some noise in the history, this PR can remove it and #605 add it back in.
